lazyscribe/experiment.py

```diff
@@ -438,6 +441,80 @@ def to_tabular(self) -> dict:
+        exist on disk already, this function will copy the artifact from the experiment
+        directory to the repository, increment the version, and call
+        :py:meth:`lazyscribe.repository.Repository.save`.
+
+        Parameters
+        ----------
+        repository : Repository
+            The :py:class:`lazyscribe.repository.Repository` to promote the artifact to.
+        name : str
+            The artifact to promote.
+
+        Raises
+        ------
+        ArtifactLogError
+            Raised if the artifact to be promoted is older than the latest version available in the
+            repository.
+        ArtifactLoadError
+            Raised if there is no artifact with the name ``name`` in the experiment.
+        """
+        for artifact in self.artifacts:
+            if artifact.name == name:
+                try:
+                    meta_ = repository.get_artifact_metadata(name)
+                    if (
+                        datetime.strptime(meta_["created_at"], "%Y-%m-%dT%H:%M:%S")
+                        > artifact.created_at
+                    ):
+                        raise ArtifactLogError(
+                            f"Artifact `{name}` is older than the latest version available in the repository."
+                        ) from None
+                    new_handler = evolve(artifact, version=meta_["version"] + 1)
+                except ValueError:
+                    new_handler = copy(artifact)
+
+                if artifact.dirty:
+                    LOG.debug(
+                        f"Artifact '{name}' does not exist on the filesystem yet."
+                    )
+                    repository.artifacts.append(new_handler)
+                else:
+                    # The artifact is on disk, we will have to copy it over
+                    curr_path = self.path / artifact.fname
+                    new_path = repository.dir / artifact.name
+                    LOG.debug(f"Copying '{curr_path!s}' to '{new_path!s}{os.sep}'")
+                    if not self.fs.isdir(f"{new_path!s}{os.sep}"):
+                        LOG.debug(f"Creating '{new_path!s}{os.sep}'")
+                        self.fs.mkdir(f"{new_path!s}{os.sep}", create_parents=True)
+                    self.fs.copy(str(curr_path), f"{new_path!s}{os.sep}")
+
+                    repository.artifacts.append(new_handler)
+                    LOG.info(
+                        "Calling `save` on the repository since the artifact exists on disk already."
+                    )
+                    try:
+                        repository.save()
+                    except SaveError as exc:
+                        LOG.exception(exc, exc_info=True)
+                        LOG.info(
+                            f"Save failed, deleting '{(new_path / artifact.fname)!s}'..."
+                        )
+                        self.fs.rm(str(new_path / artifact.fname))
+                        del repository.artifacts[-1]

```

Comment:
I don't understand why we cover up the exception here...
Is it a normal situation?
Why can it happen? (In addition to a disk failure.)
We just remove the artifact and the user's (programmer's) code doesn't even know it...

Reply:
I think it will be rare, but the motivation here is to prevent a situation where the filesystem and the `repository.json` are out of sync. If there is some failure with the `repository.save` call, we could have a dangling artifact file with no corresponding entry in the JSON.

We aren't removing the artifact from the experiment, just the copied version that we were moving as part of the promotion process.

Reply:
@ak-gupta , I meant why we _hide_ the exception. Thought was to re-raise it (instead of/in addition to the warning). IMHO if we hide the exception, the user is exactly able get out of sync (as he doesn't get an error (i.e, exception)).

Reply:
Ah, sorry for the miscommunication. We can re-raise the error.